### ISSUE TITLE
apoc.export.cypher.query writes out invalid cypher

### ISSUE THREAD

**Issue by MarkPAndrews** [neo4j-contrib/neo4j-apoc-procedures#1209]
## Guidelines

Please note that GitHub issues are only meant for bug reports/feature requests. If you have questions on how to use APOC, please ask on [the Neo4j Discussion Forum](https://community.neo4j.com/c/neo4j-graph-platform/procedures-apoc) instead of creating an issue here.

## Expected Behavior (Mandatory)
The following query matches about 10k nodes and 32k relationships (graph can be provide on request) and writes out a Cypher file.  The expectation is that the file is runable in a cypher-shell.  
```
call apoc.export.cypher.query('MATCH (t:TestData)-[r]->(:TestData) return distinct t,r',
'/tmp/test.cypher',{format:'plain'});
```
## Actual Behavior (Mandatory)
The file that is written out has 2 different syntactical errors and 1 clean-up issue
1.)  400 occurrences of an extraneous comma within an unwind statement
   i.e. `UNWIND [{Key1:"148177", properties:{prop1:" ", prop2:"2", prop3:"50.0%", }}] as row`
2.) 2057 occurrences of a reference to an _id property that doesn't exist
  ```
UNWIND [{start: {nodeId:"323"}, end: {otherId:"177"}, properties:{}}] as row
MATCH (start:Node{nodeId: row.start.nodeId, `UNIQUE IMPORT ID`: row.start._id})
MATCH (end:OtherNode{otherId: row.end.materialSupplierCode, `UNIQUE IMPORT ID`: row.end._id})
CREATE (start)-[r:IS_RELATED]->(end) SET r += row.properties;
```
3.) Some  `UNIQUE IMPORT ID`s are left behind.  Adding the following line to the files removes the remaining.  Add once per 20,000 nodes
`MATCH (n) Where exists(n.`UNIQUE IMPORT ID`)  WITH n LIMIT 20000 REMOVE n.`UNIQUE IMPORT ID`
## How to Reproduce the Problem
Run apoc.apoc.export.cypher.query
### Simple Dataset (where it's possibile)
The following dataset reproduces the missing row.start._id value.
CREATE CONSTRAINT ON (node:Node1) ASSERT (node.Id) IS UNIQUE;
CREATE CONSTRAINT ON (node:Node2) ASSERT (node.Id) IS UNIQUE;
CREATE CONSTRAINT ON (node:Node3) ASSERT (node.Id) IS UNIQUE;
CREATE CONSTRAINT ON (node:Node4) ASSERT (node.Id) IS UNIQUE;
Create (:TestData:Node1:Node2{Id:1})-[:RELATES_TO]->(b:TestData:Node3:Node4{Id:1});

### Steps (Mandatory)

  1.Run apoc.export.cypher.query above
  2.Search resulting file for ". }" and ", `UNIQUE IMPORT ID`: row.start._id"
  

## Specifications (Mandatory)

Currently used versions

### Versions

  - OS: Windows 10 - running in Ubuntu shell version 4.4.0-17763-Microsoft #379-Microsoft Wed Mar 06 19:16:00 PST 2019 x86_64 x86_64 x86_64 GNU/Linux
     -Ubuntu 18.04.1 LTS
  - Neo4j: 3.5.2 Enterprise
  - Neo4j-Apoc: 3.5.03

**Comment by conker84** [comment]
Hi @MarkPAndrews and thanks for reaching us.
I'm not able to reproduce the bug. Can you please send me the dataset?
`andrea.santurbano at larus-ba.it`
Thanks a lot!

**Comment by conker84** [comment]
@MarkPAndrews ping :)

**Comment by MarkPAndrews** [comment]
Sorry, I'm not getting notifications from Git for some reason.  I've send the dataset to @jexp.  I hesitate to share it widely since it has sensitive client data in it.  I'll email Michael.

**Comment by conker84** [comment]
@MarkPAndrews I think I did some progress, but without the dataset I'm not able to reproduce the problem. Btw I hope the following jar could help:
[apoc-3.5.0.4-all.jar.zip](https://github.com/neo4j-contrib/neo4j-apoc-procedures/files/3326276/apoc-3.5.0.4-all.jar.zip)

Looking forward to your feedback

**Comment by conker84** [comment]
@MarkPAndrews ping :)

**Comment by MarkPAndrews** [comment]
Thanks for the reminder.  I got back to cleaning up the data but after renaming all of the nodes and relationships and scrambling names and contact information the problem not longer happens.
I did try the new jar that you sent.  It appears to have fixed the  ". }" issue but the missing "row.start._id" is still a problem.  
I plan to keep trying to reproduce this problem with cleansed data.  I'll let you know when I have something I can share.

**Comment by MarkPAndrews** [comment]
@conker84 One note on the new jar file.  If I run this cypher:
"call apoc.export.cypher.all('/tmp/myfile.cypher',{format:'plain'});"
It seems to have changed the file location from being absolute ("/tmp/myfile.cypher") to a sub directory of the import directory ("/neo4j/import/tmp/myfile.cypher").
Not sure if that was intentional.

**Comment by MarkPAndrews** [comment]
@conker84 I've updated the description with a very simple dataset that reproduces the problem with the missing `row.start._id`.  I was able to attach a debugger and found the problem but I'll leave to more qualified hand to fix it. :)
The problem is cause by the nodes having at least 1 label that doesn't a unique constraint.  In the `AbstractCypherFormatter` calls the `writeRelationshipNodeIds` methods finds at least one Id with a unique constraint so it doesn't write out the `_id` property.  But `getNodeIdProperties` in `CypherFormatterUtils` sees that one of the Labels doesn't have a unique constraint (line 61) so it adds the `UNIQUE IMPORT ID` property for the MATCH clause.
It seems like it would be safe for `getNodeIdProperties` to only add the `UNIQUE IMPORT ID` if none of the labels.

A simple fix would be to change the assignment of `uniqueLabelFound` to something like this:
`uniqueLabelFound = uniqueLabelFound  || isUniqueLabelFound(node, uniqueConstraints, labelName);`
But I see that assignment logic in a lot of places - so again, I'll leave it to more capable hands.
